src/cc/frontends/p4/compiler/ebpfAction.py

```diff
@@ -0,0 +1,327 @@
+from p4_hlir.hlir import p4_action, p4_field, p4_signature_ref, p4_header_instance
+import ebpfProgram
+from programSerializer import ProgramSerializer
+from compilationException import *
+import ebpfScalarType
+import ebpfCounter
+import ebpfType
+import ebpfInstance
+
+
+class EbpfActionData:
+    def __init__(self, name, argtype):
+        self.name = name
+        self.argtype = argtype
+
+
+class EbpfActionBase:
+    def __init__(self, p4action):
+        self.name = p4action.name
+        self.hliraction = p4action
+        self.builtin = False
+        self.arguments = []
+
+    def serializeArgumentsAsStruct(self, serializer):
+        serializer.emitIndent()
+        serializer.appendFormat("/* no arguments for {0} */", self.name)
+        serializer.newline()
+
+    def serializeBody(self, serializer, valueName, program):
+        serializer.emitIndent()
+        serializer.appendFormat("/* no body for {0} */", self.name)
+        serializer.newline()
+
+    def __str__(self):
+        return "EbpfAction({0})".format(self.name)
+
+
+class EbpfAction(EbpfActionBase):
+    unsupported = [
+        # The following cannot be done in EBPF
+        "add_header", "remove_header", "execute_meter", "clone_ingress_pkt_to_egress",
+        "clone_egress_pkt_to_egress", "generate_digest", "resubmit",
+        "modify_field_with_hash_based_offset", "truncate", "push", "pop",
+        # The following could be done, but are not yet implemented
+        # The situation with copy_header is complicated, because we don't do checksums
+        "copy_header", "count",
+        "register_read", "register_write"]
+
+    # noinspection PyUnresolvedReferences
+    def __init__(self, p4action, program):
+        EbpfActionBase.__init__(self, p4action)
+        assert isinstance(p4action, p4_action)
+        assert isinstance(program, ebpfProgram.EbpfProgram)
+
+        self.builtin = False
+        self.invalid = False  # a leaf action which is never called from a table can be invalid.
+
+        for i in range(0, len(p4action.signature)):
+            param = p4action.signature[i]
+            width = p4action.signature_widths[i]
+            if width is None:
+                self.invalid = True
+                return
+            argtype = ebpfScalarType.EbpfScalarType(p4action, width, False, program.config)
+            actionData = EbpfActionData(param, argtype)
+            self.arguments.append(actionData)
+
+    def serializeArgumentsAsStruct(self, serializer):
+        if self.invalid:
+            raise CompilationException(True, "{0} Attempting to generate code for an invalid action", self.hliraction)
```

Comment:
Though we're not enforcing anything in the code as yet (pylint, etc.), it would be good to keep lines under 80 characters if possible. This isn't my favorite python coding guideline, but it seems to be the general practice.